runtime/bin/snapshot_utils.cc

```diff
@@ -238,9 +239,60 @@ static AppSnapshot* TryReadAppSnapshotElf(
   }
   return new ElfAppSnapshot(handle, vm_data_buffer, vm_instructions_buffer,
                             isolate_data_buffer, isolate_instructions_buffer);
-  return nullptr;
 }
 
+class MachOAppSnapshot : public AppSnapshot {
+ public:
+  MachOAppSnapshot(Dart_LoadedMachO* mach_o,
+                 const uint8_t* vm_snapshot_data,
+                 const uint8_t* vm_snapshot_instructions,
+                 const uint8_t* isolate_snapshot_data,
+                 const uint8_t* isolate_snapshot_instructions)
+      : mach_o_(mach_o),
+        vm_snapshot_data_(vm_snapshot_data),
+        vm_snapshot_instructions_(vm_snapshot_instructions),
+        isolate_snapshot_data_(isolate_snapshot_data),
+        isolate_snapshot_instructions_(isolate_snapshot_instructions) {}
+
+  virtual ~MachOAppSnapshot() { Dart_UnloadMachO(mach_o_); }
+
+  void SetBuffers(const uint8_t** vm_data_buffer,
+                  const uint8_t** vm_instructions_buffer,
+                  const uint8_t** isolate_data_buffer,
+                  const uint8_t** isolate_instructions_buffer) {
+    *vm_data_buffer = vm_snapshot_data_;
+    *vm_instructions_buffer = vm_snapshot_instructions_;
+    *isolate_data_buffer = isolate_snapshot_data_;
+    *isolate_instructions_buffer = isolate_snapshot_instructions_;
+  }
+
+ private:
+  Dart_LoadedMachO* mach_o_;
+  const uint8_t* vm_snapshot_data_;
+  const uint8_t* vm_snapshot_instructions_;
+  const uint8_t* isolate_snapshot_data_;
+  const uint8_t* isolate_snapshot_instructions_;
+};
+
+static AppSnapshot* TryReadAppSnapshotMachO(
+    const char* script_name) {
+  const char* error = nullptr;
+  const uint8_t *vm_data_buffer = nullptr, *vm_instructions_buffer = nullptr,
+                *isolate_data_buffer = nullptr,
+                *isolate_instructions_buffer = nullptr;
+  Dart_LoadedMachO* handle = nullptr;
+  handle = Dart_LoadMachO(script_name, &error, &vm_data_buffer,
+                          &vm_instructions_buffer, &isolate_data_buffer,
+                          &isolate_instructions_buffer);
+  if (handle == nullptr) {
+    Syslog::PrintErr("Loading failed: %s\n", error);
```

Comment:
It's possible this should be removed on landing. I I don't think the elf pathway logs here and this likely causes a spurious log during the "fall back to elf" pathway.